### PR TITLE
Refactor dynamic 2

### PR DESCRIPTION
**Pull request by rjpolackwich**
This is the follow up PR to refactor-dynamic. The main objective of that work was the reduce complexity and redundancy in meta.py and establish optimized, well defined patterns for subclassing and initiation. Includes various method bug fixes. Also happens to include some work in MultiCurlFetch that implements configurable worker pool queues for throttling the driver as well as retry capability. 

The major take away from refactor-dynamic is the simplification of meta.py. There is now only one class, GeoDaskImage, which subclasses DaskImage and is intended to be used for subclassing custom image classes. The function of the DaskMeta object now is to act as an adaptor for the init params passed to GeoDaskImage, giving us more flexibility in the structures we can pass to init (we should be able to init w/ a dict, other dask arrays, etc without forcing a type constraint) as well as a place to easily adapt new structures. And anyway, all the main first class objects in this API are themselves dask arrays, and inherently carry the metadata and manage it as necessary, so we should just refer directly to it as needed at the instantiation process or whatever. 

Summary + other noteworthy refactor-dynamic things:
* We have one main class, `GeoDaskImage`, that can be initiated on its own or subclassed.
* `DaskMeta` is less critical and more flexible, adaptor-like role
* As a result, the `DaskImage.create` pattern has been removed along with the DaskMeta infection hook used there via dask plugins. The new creation pattern is accomplished via normal instantiation or subclassing and potentially implementing `__new__` if necessary.
* Useage of metaclasses and subclasshooks removed
* `DaskGeoImage.__getattr__` custom wrapper removed. Not sure what the intended purpose of this was, looked to me like it was meant to ensure that dask array functions that return dask arrays were sure to return objects that look like GeoImage by manually altering the returned obj attrs and methods. This is unnecessary because da.Array can be subclassed as usual and returned objects will derive from the subclass as expected according to what the array func returns. The removal of this wrapper fn has immediate and pleasant consequences: in the jupyter notebook,  doc/code inspection via `??` will show us the actual relevant code, and stacktraces are far more linear and easy to understand as opposed to every method call being sandwiched by that wrapper all the way down the inheritance chain.  
* The functionality of `__get_item__` in terms of geo stuff is all handled in one place on `GeoDaskImage` and is "self-contained" 
* `GeoDaskImage.__get_item__` now handles indexing with ellipses, which we needed to be able to support. It might not do it right for images of dimsize != 3 though I have to check.

The goal of refactor-dynamic-2 is two-fold, the primary objective being the engineering and implementation of a better workflow (and codebase) for developing image/product sources while the secondary objective aims to structure gbdxtools to a point where DaskImage/GeoDaskImage/etc are closer to becoming independent of Image Products/Driver-classes/ Ipe/RDA/etc. 

The result of objective 1 is the image driver concept. I settled on this because i think that image products from diff source providers will need to follow the same kind of  a) configure the params the that will determine/define the image prodcut via "asset type" or "ipe graph" or whatever which will often involve some kind of input option parsing w/ predetermined default support and then b) use those config params in a set of functions that actually build the dask graph needed for GeoDaskImage. 
One v potentially useful feature of the RDADriver classes is the built in option-parser thanks to metaclass pattern. Drivers specify a set of supported options and default options that can be used to parse and set image options at initialization. The nice thing is that Driver classes can be subclassed to easily implement small or large variation in option support without explicitly defining a whole new key map, but maintaining transparency through the class attributes. They can be used to customize image products on creation as well, according to class or whatever.  This could be handy, eg, when a large number of image classes support a large set of product specs but with variation in option support and default values. Image classes themselves define the critical functions used to build the dask graph, and so can be used to initiate GeoDaskImage without the use of drivers. 

I think the driver mechanism is useful conceptually as well because it takes an image class and some input params and creates a rather arbitrary object that's not native to the concept of an image class, imo. 

The result of the second objective is the general deconstruction of the IpeImage class for the standard RDABaseImage interface and the introduction of the `GeoAdapter` class. The motivation behind GeoAdapter is the appeal of the concept of a "standardized geo interface" that we could implement as part of GeoDaskImage or something. If we could do something like that, DaskGeoImage could stand on its own and methods could be developed without worrying about metadata formatting or whatever.

TLDR; This PR establishes new patterns for source image class GeoDaskImage adaption via driver objects responsible for the burden of dask-graph creation from a set of user inputs. It also introduces the concept of a GeoAdapter class, which could be used to make the entire GeoDaskArray class truly geo-metadata-format agnostic, which has obvious advantages.  So in concept (and application), there are two core first class objects: GeoDaskArray and the Source Image class. The Utopian Future 
 would have developers writing Image classes where they can concentrate on doing things like adding custom sensor or geo based methods as necessary and then relying on the existing tooling libraries (ImageDrivers, GeoAdapters, or subclasses thereof) to support GeoDaskImage, which runs on a declared interface independent of metadata formatting language. 

Going forward, Open questions, I appreciate your comments on:
* The implementation of the Driver class (see RDABaseImage) is kind of unusual, maybe could be better?
* Image classes do not define their option support but rely on their drivers - this seems weird to me but couldn't get around it. The image instances, though, do have an `options` and `__default_options__` attr that are set by the driver and can be useful to api users.
* The GeoAdapter class... useful? I think maybe but I'm doing it wrong so far
* RDABaseImage should take the place of IpeImage and implement the "post creation hooks" in one place
* DaskImage doesn't provide a whole lot of methodologies. Should it be combined with GeoDaskImage, or is it useful as a separate class? I can imagine more methods that could be added there, esp optimizations for ndarray functions kinda thing

### CODE REVIEW

**Comment by chelm** [comment]
Nice at first glance this looks really good. Going to keep reviewing it.

**Review by chelm — APPROVED** [review]
Its a lot to comb through, but nothing jumps out besides it being a really great improvement in several places. Nice work. 

My vote is we merge it into dev and make a push to get a few super sketchy areas (fetching mostly) tested in the unittests.

**Comment by chelm** [comment]
Great write up @rjpolackwich. 

My thoughts on the "open questions":

`The implementation of the Driver class (see RDABaseImage) is kind of unusual, maybe could be better?`

Good enough for, a bit funky but its the key to understanding the inheritance...

`Image classes do not define their option support but rely on their drivers - this seems weird to me but couldn't get around it. The image instances, though, do have an options and __default_options__ attr that are set by the driver and can be useful to api users.`

I like putting more into the drivers, so this make sense.

`The GeoAdapter class... useful? I think maybe but I'm doing it wrong so far`

Seems like we should go all in on organizing every "geo" related method or prop into a single "Geo Adapter"

`RDABaseImage should take the place of IpeImage and implement the "post creation hooks" in one place`

Sure. 

`DaskImage doesn't provide a whole lot of methodologies. Should it be combined with GeoDaskImage, or is it useful as a separate class? I can imagine more methods that could be added there, esp optimizations for ndarray functions kinda thing`

I think the original motivation for having geo and non-geo split up still holds. In theory there could be several things we want inside DaskImage that are unrelated to having anything to do with Geo. So im a fan of keeping them as is for now.

**Comment by chelm** [comment]
So should we merge it @rjpolackwich or do you want to see if we can @pramukta's eyes on it. I think its worth mergin

**Comment by pramukta** [comment]
Nice work @rjpolackwich!  I have to admit this is hard to review as a whole.  However, to a large degree the changes warrant that style PR.  I think one of the most important pieces is that our legacy adaptation of older dask version patterns is more or less gone, so the pain of tracking the upstream dep is less.  I'm going to start with more conceptual stuff because of the complexity here.  Also I'm lazy.

A couple of comments on your concerns about drivers.  This is because of upstream interface inconsistencies.  You have to do this whether you call it a driver or not.  We were originally hoping that RDA would normalize all of that for us, and while it does to some degree, there are still a variety of special cases.  Kinda like VectorServices where you can absolutely query for anything that you already know exists.

Which brings up the next thing which is the other valuable though aspirational bit as related to RDA.  Both the client side dask work and the RDA work are delayed compute graphs (just like Tensorflow and fellow deeep learning friends).  The dask ones are executed client side, the rda ones server side. Aspiring to a functional metadata normalization (and enabled compute) with RDA in the context a standards initiative would be an important step towards making the transition of compute from client to server seamless.  If we presented a good case, I bet cholmes could help make the community side of that happen.The necessity of the driver construct is a reflection that this is being ignored.  

Re DaskImage.  While there is a balance between features and dependencies (and installation trouble).  I do think that, for example, image visualization (esp for large images) is not super slick in matplotlib/jupyter.  If we take the juptyer notebook emulating mathematica, and jupyter lab emulating matlab as a pattern, then we ought to see desire for richer interfaces around plots (beyond what the QT backends provide).  Hell even IDL had iTools a decade ago.  This could be a good spot to prototype some of those, despite the fact that it would be better upstream.

Anyway, will do some functional tests tomorrow morning and approve after that.  Oh wait, my creds are no longer valid, if @billfgreer managed to get that worked out will do the tests. May want to see about getting some testing numbers up so @nricklin is happy.
